### DISCUSSION TITLE
Database security

### DISCUSSION THREAD

**Discussion by christrunk** [supabase/supabase#3634] [Questions] [answered]
Hi there,

I've worked in or managed development teams for almost 15 years. In whatever company or system I've worked on, the database is securely tucked away behind a firewall and/or multiple layers of authentication.

On Supabase, the Postgres DB seems to be fully accessible with read/write with just a plain username and password.

Is that right?

Could it be vulnerable to a stolen password, or brute-forcing?

I'm not a Postgres or DB guy, just the part of the team that would be worried about security, so I might be misunderstanding the setup with Supabase.

Can someone help me sleep better at night please? 😀

**Comment by dshukertjr** [comment, marked as answer]
Hi @christrunk!

Postgres DB used in Supabase is behind an authentication layer just like any other databases out there.

As long as you have configured [row level security policies](https://supabase.io/docs/guides/auth/row-level-security), your database should be nice and secure. Gotrue, the authentication server used in Supabase has mechanisms to prevent brute-forcing as you can see [here](https://github.com/supabase/gotrue/blob/master/api/api.go#L125). Whether a stolen password could be critical or not is up to how you configure your security policy, I think.

**Reply by christrunk** [reply]
That's really helpful, thanks @dshukertjr !

**Reply by dinvlad** [reply]
Could you please elaborate, does `Gotrue` handle _Postgres_ connection auth also? I thought it's only used by the API proxy for Postg*REST* etc.?

Raw database port (even with SSL encryption) exposed to the world usually pops up as a High finding on external security audits required for compliance. So I would imagine firewalling it would be quite a required feature for SOC2 etc..

**Reply by dshukertjr** [reply]
Hi @dinvlad 

>I thought it's only used by the API proxy for PostgREST etc.?

This is correct. Gotrue only handles auth used by the API proxy for PostRest and realtime.

>Raw database port (even with SSL encryption) exposed to the world usually pops up as a High finding on external security audits required for compliance. So I would imagine firewalling it would be quite a required feature for SOC2 etc..

Everything that Supabase offers sits behind a API gateway platform called Kong, so we do not expose the raw database port to the public actually. You can read more about it on the ['How It Works'](https://github.com/supabase/supabase#how-it-works) section on Supabase's readme!

**Reply by dinvlad** [reply]
Thanks for the link - it's a bit unclear how that applies to the raw database connection though, since I thought Kong only proxies the REST level. So if I connect to `postgres://postgres:<password>@db.xxxyyyzzz.supabase.co:5432/postgres`, are you saying that the port `5432` connection also goes via Kong? And in that scenario, since it's a "raw" PostgreSQL protocol connection (with or without SSL), what features of Kong make it "safer" to connect from `0.0.0.0` via the proxy Kong port `5432` to the internal Postgres port `5432`?

(Also, looking at Docker configuration for Kong [here](https://github.com/supabase/supabase/blob/64afd803e60b34f6516e9e5c8041084766a91270/docker/docker-compose.yml#L25-L39), it only appears to expose ports `8000`/`8443`, as would be expected for HTTP(S))

Sorry for challenging you with these questions - as a security professional, I'm just naturally in the mindset of "trust but verify", so any more links to the code/configuration/threat model you could share on this would be really helpful!

**Reply by dshukertjr** [reply]
@dinvlad I probably am not qualified to answer this question, so I have tagged someone from the infrastructure team 😂 He should be able to clarify your concerns. Sorry for taking some time to get back.

**Reply by dinvlad** [reply]
Thanks, much appreciated!

**Reply by burmecia** [reply]
hi @dinvlad , yes you're right, that is a direct database connection which uses raw postgres protocol and can be used by `psql` command. Kong just builds another layer on top of it and provides restful api to client. As for the security concern, it is true that postgres port 5432 is open to public, but its password is randomly generated and ssl is used by default in most tools such as psql. Although you can intensionally disable ssl when using psql but I don't think most people will do that. So it is mostly secure as long as the user follows common security best practice, such like saving password safely, using ssl, rotating password regularly and etc.

**Reply by dinvlad** [reply]
Thank you for looking into it, and providing more explanation. Are there any plans to add database firewalling support for those customers who're more concerned about it? I believe this will target a larger proportion of customers, especially the ones dealing with sensitive/under-compliance data.

**Reply by christrunk** [reply]
@burmecia  - Am I right in saying that there's at least fail2ban sitting in front of the direct database connection, protecting against brute force attempts?

**Reply by darora** [reply]
While we can't talk about the specifics of all the measures we use to secure our databases, we do have mechanisms in place to protect against brute-force attacks.

Later this year, we'll likely be making available additional network isolation features (e.g. IP whitelisting, private tunnels); we don't have a firm timeline on these as of yet.

**Reply by dinvlad** [reply]
OK, thanks for your attention to it!

**Reply by christrunk** [reply]
> Later this year, we'll likely be making available additional network isolation features (e.g. IP whitelisting, private tunnels); we don't have a firm timeline on these as of yet.

Awesome, that's the sort of feature I was expecting TBH. Glad you're on it and understand why we might have to wait a while too.

**Reply by suhjohn** [reply]
Hi @darora , is there any update on this? It's the one thing blocking me from using Supabase.

**Reply by davidbehan** [reply]
Interested in this feature too. Any update or timeline yet? Thanks

**Reply by rorito** [reply]
This is also our biggest blocker preventing us from adopting Supabase, is there an update on network isolation features timeline?

**Reply by siquick** [reply]
Any updates? It’s a vital feature for any serious application

**Reply by darora** [reply]
IP restrictions are currently being tested in beta. Please reach out to Support via the dashboard if you'd like to have it enabled for your project.

**Reply by rr404** [reply]
Hello,

I work at CrowdSec, and we recently released the first version of protection scenarios tailored to Supabase:
https://www.crowdsec.net/blog/hardening-self-hosted-supabase

I'd gladly guide you through the process if you need help.
And eager to get some feedback and requests for more attacks or misbehavior scenarios you'd like to get or create.

It's OpenSource so we're always happy when people join the community and participate to improve everybody's protection against malicious IPs.

**Reply by licarth** [reply]
Hi, thanks for Supabase, which seems to be a really great tool.

I was a bit shocked as well when I saw that the DB was publicly exposed. I'm really looking forward having some TCP proxy to access the DB being authed as a Supabase user. I'm thinking of something like Google's [cloud-sql-proxy](https://cloud.google.com/sql/docs/mysql/sql-proxy).

Would be great to be able to use this in the CI for our database migrations, without having to rotate passwords. Something like `supabase db proxy -p 5432` that requires either a Supabase token or to be logged as a user that has read/write access to the database.

Cheers,

**Reply by sweatybridge** [reply]
> Something like supabase db proxy -p 5432 that requires either a Supabase token or to be logged as a user that has read/write access to the database.

I can look into supporting this through the CLI. We currently expose a `/v1/projects/{ref}/database/query` [API](https://api.supabase.com/api/v1#/projects%20(beta)/v1RunQuery) that serves as http proxy for running arbitrary SQL statements. The CLI can use that instead of connecting natively through the postgres port.

One limitation of this approach is that queries run through proxy is restricted to a timeout of 15 seconds. In the long run, we hope to support authentication with PAT through native postgres connection. That would require new features to be built into [supavisor](https://github.com/supabase/supavisor).
